### PR TITLE
MRG: duality gap for TF-MxNE

### PR DESCRIPTION
**Pull request by joewalter**

### CODE REVIEW

**Comment by agramfort** [comment]
thanks @joewalter !

@mathurinm this one is for you :)

**Comment by agramfort** [comment]
and for @yousrabk too :)

File: mne/inverse_sparse/mxne_optim.py [hunk 1]
```diff
@@ -714,26 +714,141 @@ def norm_l1_tf(Z, shape, n_orient):
...
+               n_orient, highest_d_obj, screening):
+    """Duality gaps for the time-frequency mixed norm inverse problem
+
+    For details see:
+    Wang J., and Ye J.,
+    Two-layer feature reduction for sparse-group lasso via decomposition of
+    convex sets, in Advances in Neural Information Processing Systems (NIPS),
+    vol. 27, pp. 2132-2140, 2014.
+
+    Bach F., Jenatton R., Mairal J., and Obozinski G.,
```
Comment (mathurinm) [inline mne/inverse_sparse/mxne_optim.py]: I don't see the dual of Sparse group Lasso being derived in this

File: mne/inverse_sparse/mxne_optim.py [hunk 2]
```diff
@@ -714,26 +714,141 @@ def norm_l1_tf(Z, shape, n_orient):
...
+        Y[o::n_orient] *= shrink
+    return Y
+
+
+def dgap_l21l1(M, G, Z, active_set, alpha_space, alpha_time, phi, phiT, shape,
+               n_orient, highest_d_obj, screening):
+    """Duality gaps for the time-frequency mixed norm inverse problem
+
+    For details see:
+    Wang J., and Ye J.,
```
Comment (mathurinm) [inline mne/inverse_sparse/mxne_optim.py]: Maybe we could even point to Thm5 of this ? It took me a bit of time to find the expression of the dual feasible that we use

File: mne/inverse_sparse/mxne_optim.py [hunk 3]
```diff
@@ -714,26 +714,141 @@ def norm_l1_tf(Z, shape, n_orient):
...
+    if n_orient > 1:
+        active_set = np.tile(active_set[:, None], [1, n_orient]).ravel()
+    for o in range(n_orient):
+        Y[o::n_orient] *= shrink
+    return Y
+
+
+def dgap_l21l1(M, G, Z, active_set, alpha_space, alpha_time, phi, phiT, shape,
+               n_orient, highest_d_obj, screening):
+    """Duality gaps for the time-frequency mixed norm inverse problem
```
Comment (mathurinm) [inline mne/inverse_sparse/mxne_optim.py]: gap

File: mne/inverse_sparse/mxne_optim.py [hunk 4]
```diff
@@ -714,26 +714,141 @@ def norm_l1_tf(Z, shape, n_orient):
...
+    in Machine Learning, vol. 4, no. 1, pp. 1-106, 2012.
+
+    Gramfort A., Kowalski M., and Hamalainen, M.,
+    Mixed-norm estimates for the M/EEG inverse problem using accelerated
+    gradient methods, Physics in Medicine and Biology, 2012
+    http://dx.doi.org/10.1088/0031-9155/57/7/1937
+
+    Parameters
+    ----------
+    M : array of shape [n_sensors, n_times]
```
Comment (mathurinm) [inline mne/inverse_sparse/mxne_optim.py]: shape with ()

File: mne/inverse_sparse/mxne_optim.py [hunk 5]
```diff
@@ -714,26 +714,141 @@ def norm_l1_tf(Z, shape, n_orient):
...
+        The data.
+    G : array of shape [n_sensors, n_active]
```
Comment (mathurinm) [inline mne/inverse_sparse/mxne_optim.py]: ditto

File: mne/inverse_sparse/mxne_optim.py [hunk 6]
```diff
@@ -714,26 +714,141 @@ def norm_l1_tf(Z, shape, n_orient):
...
+        Gain matrix a.k.a. lead field.
+    Z : array of shape [n_active, n_coefs]
```
Comment (mathurinm) [inline mne/inverse_sparse/mxne_optim.py]: ditto

File: mne/inverse_sparse/mxne_optim.py [hunk 7]
```diff
@@ -714,26 +714,141 @@ def norm_l1_tf(Z, shape, n_orient):
...
+        Sources in TF domain.
+    active_set : array of bool
+        Mask of active sources.
+    alpha_space : float
+        The spatial regularization parameter. It should be between 0 and 100.
```
Comment (mathurinm) [inline mne/inverse_sparse/mxne_optim.py]: in this function nothing requires that alpha is between 0 and 100 so I'd remove the second sentence

File: mne/inverse_sparse/mxne_optim.py [hunk 8]
```diff
@@ -714,26 +714,141 @@ def norm_l1_tf(Z, shape, n_orient):
...
+    phi : instance of _Phi
+        The TF operator.
+    phiT : instance of _PhiT
+        The transpose of the TF operator.
+    shape : tuple
+        Shape of TF coefficients matrix.
+    n_orient : int
+        Number of dipoles per locations (typically 1 or 3).
+    highest_d_obj : float
+        The highest value of the dual cost so far.
```
Comment (mathurinm) [inline mne/inverse_sparse/mxne_optim.py]: I'm not fond of dual *cost* since we try to maximize it, could we say dual objective?

File: mne/inverse_sparse/mxne_optim.py [hunk 9]
```diff
@@ -714,26 +714,141 @@ def norm_l1_tf(Z, shape, n_orient):
...
+    screening : array of bool
+        Mask of screened-out sources by dynamic screening.
```
Comment (mathurinm) [inline mne/inverse_sparse/mxne_optim.py]: if this is dynamic *gap safe* screening, maybe we should point to https://arxiv.org/abs/1602.06225

File: mne/inverse_sparse/mxne_optim.py [hunk 10]
```diff
@@ -714,26 +714,141 @@ def norm_l1_tf(Z, shape, n_orient):
...
+
+    Returns
+    -------
+    gap : float
+        Dual gap
+    p_obj : float
+        Primal cost
+    d_obj : float
+        Dual cost. gap = pobj - dobj
```
Comment (mathurinm) [inline mne/inverse_sparse/mxne_optim.py]: p_obj and d_obj

File: mne/inverse_sparse/mxne_optim.py [hunk 11]
```diff
@@ -714,26 +714,141 @@ def norm_l1_tf(Z, shape, n_orient):
...
+        The highest value of the dual cost so far.
+    screening : array of bool
+        Mask of screened-out sources by dynamic screening.
+
+    Returns
+    -------
+    gap : float
+        Dual gap
+    p_obj : float
+        Primal cost
```
Comment (mathurinm) [inline mne/inverse_sparse/mxne_optim.py]: again not fond of cost (for primal it's ok but unharmonic w variable name)

File: mne/inverse_sparse/mxne_optim.py [hunk 12]
```diff
@@ -714,26 +714,141 @@ def norm_l1_tf(Z, shape, n_orient):
...
+    d_obj : float
+        Dual cost. gap = pobj - dobj
+    R : array of shape [n_sensors, n_times]
```
Comment (mathurinm) [inline mne/inverse_sparse/mxne_optim.py]: shape with ()

File: mne/inverse_sparse/mxne_optim.py (hunk 1 above)
Comment (joewalter) [inline mne/inverse_sparse/mxne_optim.py]: Ok, I'll remove this one. It is more a general reference for sparse regression and duality gap

Comment (Eric89GXL) [inline mne/inverse_sparse/mxne_optim.py]: these should go in a `References` section with `.. [1]` later and refer to them with `[1]_` somewhere here (see how we do it elsewhere)

**Comment by joewalter** [comment]
Closing for now, will reopen later

**Comment by agramfort** [comment]
it does not hurt to keep things open for weeks.

at least we don't forget it's there

**Comment by mathurinm** [comment]
The construction of the dual feasible point looks correct to me, but it's not tight (it does converge to the optimal dual point) so it's probably possible to get a lower duality gap, at the price of a more complicated formula.
It's ok for me to merge (just doing a few more comments on the code), then we can move to screening and then try another dual point in successive PRs.

File: mne/inverse_sparse/mxne_optim.py [hunk 13]
```diff
@@ -714,26 +714,141 @@ def norm_l1_tf(Z, shape, n_orient):
     return l1_norm
 
 
-@verbose
-def _tf_mixed_norm_solver_bcd_(M, G, Z, active_set, alpha_space, alpha_time,
-                               lipschitz_constant, phi, phiT,
-                               wsize=64, tstep=4, n_orient=1,
-                               maxit=200, tol=1e-8, log_objective=True,
-                               perc=None, verbose=None):
+def prox_l1_full(Y, alpha, n_orient):
```
Comment (mathurinm) [inline mne/inverse_sparse/mxne_optim.py]: I'm not really comfortable with this name, to me prox_l1 is the soft thresholding (which this function is not). I'd call this block_soft_thresholding. @agramfort your take on this?

File: mne/inverse_sparse/mxne_optim.py [hunk 14]
```diff
@@ -714,26 +714,141 @@ def norm_l1_tf(Z, shape, n_orient):
...
+    """proximity operator for l1 norm with multiple orientation support
```
Comment (mathurinm) [inline mne/inverse_sparse/mxne_optim.py]: for l2/l1 norm or l21 (mixed) norm

File: mne/inverse_sparse/mxne_optim.py [hunk 15]
```diff
@@ -714,26 +714,141 @@ def norm_l1_tf(Z, shape, n_orient):
...
+    GX = np.dot(G[:, active_set], X)
+    R = M - GX
+    penaltyl1 = norm_l1_tf(Z, shape, n_orient)
+    penaltyl21 = norm_l21_tf(Z, shape, n_orient)
+    nR2 = sum_squared(R)
+    p_obj = 0.5 * nR2 + alpha_space * penaltyl21 + alpha_time * penaltyl1
+
+    GRPhi = prox_l1_full(phi(np.dot(G[:, screening].T, R)), alpha_time,
+                         n_orient)
+    GRPhi = stft_norm2(GRPhi.reshape(*shape)).reshape(-1, n_orient)
```
Comment (mathurinm) [inline mne/inverse_sparse/mxne_optim.py]: this is no longer <G, R Phi> so the name is a bit misleading

File: mne/inverse_sparse/mxne_optim.py [hunk 16]
```diff
@@ -714,26 +714,141 @@ def norm_l1_tf(Z, shape, n_orient):
...
+    GRPhi = np.sqrt(GRPhi.sum(axis=1))
```
Comment (mathurinm) [inline mne/inverse_sparse/mxne_optim.py]: ditto

File: mne/inverse_sparse/mxne_optim.py [hunk 17]
```diff
@@ -714,26 +714,141 @@ def norm_l1_tf(Z, shape, n_orient):
...
+    dual_norm = np.amax(GRPhi)
+    scaling = alpha_space / dual_norm
+    scaling = min(scaling, 1.0)
```
Comment (mathurinm) [inline mne/inverse_sparse/mxne_optim.py]: since the computation of scaling is not trivial (and the references don't have the same notation), could we include a .rst somewhere detailing a bit the computation? We can use my gist

File: mne/inverse_sparse/mxne_optim.py [hunk 18]
```diff
@@ -714,26 +714,141 @@ def norm_l1_tf(Z, shape, n_orient):
...
+    d_obj = (scaling - 0.5 * (scaling ** 2)) * nR2 + scaling * np.sum(R * GX)
+    d_obj = max(d_obj, highest_d_obj)
+
+    gap = p_obj - d_obj
+    Gnorm = np.sqrt(groups_norm2(G[:, screening].copy().T, n_orient))
```
Comment (mathurinm) [inline mne/inverse_sparse/mxne_optim.py]: These can be precomputed in the solver and passed as argument to avoid recomputation for each gap

File: mne/inverse_sparse/mxne_optim.py [hunk 19]
```diff
@@ -808,164 +920,122 @@ def _tf_mixed_norm_solver_bcd_(M, G, Z, active_set, alpha_space, alpha_time,
...
                         R -= np.dot(G_j, phiT(Z[j]))
 
-                        if log_objective:
-                            val_norm_l21_tf += norm_l21_tf(
-                                Z[j], shape, n_orient)
-                            val_norm_l1_tf += norm_l1_tf(
-                                Z[j], shape, n_orient)
+        if log_objective:
+            if np.remainder(i + 1, 10) == 0:
+                Zd = np.vstack([Z_ for Z_ in list(Z.values()) if np.any(Z_)])
```
Comment (mathurinm) [inline mne/inverse_sparse/mxne_optim.py]: Not sure, is `Z.values() `order deterministic?

File: mne/inverse_sparse/mxne_optim.py [hunk 20]
```diff
@@ -714,26 +714,141 @@ def norm_l1_tf(Z, shape, n_orient):
...
+    dual_norm = np.amax(GRPhi)
+    scaling = alpha_space / dual_norm
+    scaling = min(scaling, 1.0)
+    d_obj = (scaling - 0.5 * (scaling ** 2)) * nR2 + scaling * np.sum(R * GX)
+    d_obj = max(d_obj, highest_d_obj)
+
+    gap = p_obj - d_obj
+    Gnorm = np.sqrt(groups_norm2(G[:, screening].copy().T, n_orient))
+    screening_ = (GRPhi + np.sqrt(2. * gap) * Gnorm) > alpha_space
+    screening[screening] = np.repeat(screening_, n_orient)
```
Comment (mathurinm) [inline mne/inverse_sparse/mxne_optim.py]: This is very subjective but the `screening` name might be a bit ambiguous. I read `disabled` in sklearn. Maybe `disabled_sources` (to contrast with `screening_` which would correspond to disabled_pos), or is it too long?

File: mne/inverse_sparse/mxne_optim.py [hunk 21]
```diff
@@ -808,164 +920,122 @@ def _tf_mixed_norm_solver_bcd_(M, G, Z, active_set, alpha_space, alpha_time,
                         active_set_j[:] = True
                         R -= np.dot(G_j, phiT(Z[j]))
 
-                        if log_objective:
-                            val_norm_l21_tf += norm_l21_tf(
-                                Z[j], shape, n_orient)
-                            val_norm_l1_tf += norm_l1_tf(
-                                Z[j], shape, n_orient)
+        if log_objective:
+            if np.remainder(i + 1, 10) == 0:
```
Comment (mathurinm) [inline mne/inverse_sparse/mxne_optim.py]: any reason to avoid (i + 1) % 10 == 0 ?

File: mne/inverse_sparse/mxne_optim.py (hunk 13 above)
Comment (joewalter) [inline mne/inverse_sparse/mxne_optim.py]: We called it prox_l1 also in the paper, but with a description that it is a block thresholding if free or loose source orientations are used. 
@mathurinm would adding such a description be enough?
@agramfort what's your opinion on changing the function name?

**Comment by agramfort** [comment]
I know it's a misnomer but I would keep it like this with a clear note
at the beginning of the docstring.

File: mne/inverse_sparse/mxne_optim.py (hunk 19 above)
Comment (joewalter) [inline mne/inverse_sparse/mxne_optim.py]: any idea how to test?

File: mne/inverse_sparse/mxne_optim.py [hunk 22]
```diff
@@ -48,6 +47,26 @@ def prox_l21(Y, alpha, n_orient, shape=None, is_stft=False):
...
+    shape : None | tuple
+        Shape of TF coefficients matrix.
+    is_stft : bool
+        If True, Y contains TF coefficients.
+
+    Returns
+    -------
+    Y : array, shape (n_sources, n_coefs)
+        The output data.
+    active_set : array of bool
```
Comment (agramfort) [inline mne/inverse_sparse/mxne_optim.py]: shape?

File: mne/inverse_sparse/mxne_optim.py [hunk 23]
```diff
@@ -91,10 +110,38 @@ def prox_l21(Y, alpha, n_orient, shape=None, is_stft=False):
     return Y, active_set
 
 
-def prox_l1(Y, alpha, n_orient):
+def prox_l1(Y, alpha, n_orient, remove_nonactive=True):
```
Comment (agramfort) [inline mne/inverse_sparse/mxne_optim.py]: why do you need this option?

**Comment by joewalter** [comment]
Travis fails due to time out.
@mathurinm @agramfort  can you do another round of review?
@mathurinm how can we add your finding on the dual computation? Can you condense your gist in order to add it as a note in the doc of the dual gap function, so that we don't need another file?

File: examples/inverse/plot_time_frequency_mixed_norm_inverse.py [hunk 24]
```diff
@@ -84,7 +84,7 @@
 
 # Compute TF-MxNE inverse solution
 stc, residual = tf_mixed_norm(evoked, forward, cov, alpha_space, alpha_time,
-                              loose=loose, depth=depth, maxit=200, tol=1e-4,
+                              loose=loose, depth=depth, maxit=200, tol=1e-8,
```
Comment (agramfort) [inline examples/inverse/plot_time_frequency_mixed_norm_inverse.py]: why this change?

File: mne/inverse_sparse/mxne_optim.py [hunk 25]
```diff
@@ -714,55 +768,124 @@ def norm_l1_tf(Z, shape, n_orient):
...
+
+    Returns
+    -------
+    gap : float
+        Dual gap
+    p_obj : float
+        Primal objective
+    d_obj : float
+        Dual objective. gap = p_obj - d_obj
+    R : array of shape (n_sensors, n_times)
```
Comment (agramfort) [inline mne/inverse_sparse/mxne_optim.py]: R : array, shape (n_sensors, n_times)

File: examples/inverse/plot_time_frequency_mixed_norm_inverse.py (hunk 24 above)
Comment (agramfort) [inline examples/inverse/plot_time_frequency_mixed_norm_inverse.py]: forget it. It is still super fast

**Comment by agramfort** [comment]
I played with the example plot_time_frequency_mixed_norm_inverse and it works like a charm ! nice dual gaps going down !

don't have time to look in closer detail

File: mne/inverse_sparse/mxne_optim.py [hunk 26]
```diff
@@ -714,55 +768,124 @@ def norm_l1_tf(Z, shape, n_orient):
...
+        Current residual of M - G * X
```
Comment (mathurinm) [inline mne/inverse_sparse/mxne_optim.py]: Current residual (M - G * X)

File: mne/inverse_sparse/mxne_optim.py [hunk 27]
```diff
@@ -771,237 +894,167 @@ def _tf_mixed_norm_solver_bcd_(M, G, Z, active_set, alpha_space, alpha_time,
...
-                        if log_objective:
-                            val_norm_l21_tf += norm_l21_tf(
-                                Z[j], shape, n_orient)
-                            val_norm_l1_tf += norm_l1_tf(
-                                Z[j], shape, n_orient)
+                        R -= np.dot(G_j, phiT(Z[jj]))
+
+        if log_objective:
+            if (ii + 1) % 10 == 0:
+                Zd = np.vstack([Z_ for Z_ in Z.values() if np.any(Z_)])
```
Comment (mathurinm) [inline mne/inverse_sparse/mxne_optim.py]: [Z[pos] for pos in range(n_positions) if np.any(Z[pos])] is probably a bit slower (neglectable compared to what comes next) but guarantees that the order of positions is correct.

**Comment by agramfort** [comment]
@joewalter can you rebase to make travis happy?

@mathurinm good to go from your end?

**Comment by mathurinm** [comment]
All good for me. Nice work @joewalter

**Comment by agramfort** [comment]
@joewalter please rebase to make travis happy and add a line to what's new

thanks

**Comment by joewalter** [comment]
Travis error seems unrelated to this PR. 
Setting to MRG now requesting you to review.

**Comment by agramfort** [comment]
all green ! thx @joewalter
